library/certificate_request.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+DOCUMENTATION = """
+---
+module: certificate_request
+short_description: Manage SSL/TLS certificates.
+description:
+  - The C(certificate_request) module takes a name, desired
+    certificate request attributes and certificate properties.
+  - The request is generated and sent to the CA to sign.
+
+options:
+  name:
+    description:
+      - Name of the certificate. A full path can be used
+        to choose the directory where files will be stored.
+    required: true
+  dns:
+    description:
+      - Domain (or list of domains) to be included in the
+        certificate. Also can provide the default value for
+        common_name.
+    required: false
+  ca:
+    description:
+      - CA that will issue the certificate. The available options
+        will vary depending on each provider.
+    required: true
+  provider:
+    description:
+      - The underlying method used to request and manage the
+        certificate.
+    required: false
+    default: certmonger
+author:
+  - Sergio Oliveira Campos (@seocam)
+"""
+
+EXAMPLES = """
+# Certificate for single domain
+- name: Ensure certificate exists for www.example.com
+  certificate_request:
+    name: single-example
+    dns: www.example.com
+    ca: self-sign
+
+# Certificate for multiple domains
+- name: Ensure certificate exists for multiple domains
+  certificate_request:
+    name: many-example
+    dns:
+      - www.example.com
+      - example.com
+    ca: self-sign
+"""
+
+RETURN = ""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.certificate.providers import certmonger
```

Comment:
It looks that most of the actual logic is under `module_utils.certificate.providers`. Would it make sense to have one module per provider, instead of having a single module that dispatches to the appropriate provider? That way you would not need to hardcode specific providers into common code.

Reply:
Most of the logic is in `providers.base`. Even if we have a module per provider I think that would basically remains the same.

I get your point about how the coupling was designed... 
We could use a constant in the providers.__init__ and just import this constant here. What do you think?

Reply:
Do you think this way would be better? fa8a7981490bf0ceb24f02961238f861705940c5